tests/frontier/opcodes/test_callcode.py

```diff
@@ -0,0 +1,117 @@
+"""
+abstract: Tests the CALLCODE gas consumption with a CALL/CALLCODE value transfer.
+
+    Tests an EthereumJS bug where the CALLCODE gas was incorrectly calculated from a
+    CALL value transfer: https://github.com/ethereumjs/ethereumjs-monorepo/issues/3194
+
+    Test setup: 2 smart contract accounts exist where AAAB CALL/CALLCODEs AAAA, which then
+    CALLCODEs a non-existent contract AAAC. The result of AAAA's CALL/CALLCODE is stored in AAAB's
+    storage slot 0.
+"""
+
+from typing import Dict
+
+import pytest
+
+from ethereum_test_tools import (
+    Account,
+    Environment,
+    StateTestFiller,
+    TestAddress,
+    Transaction,
+    to_address,
+)
+from ethereum_test_tools.vm.opcode import Opcodes as Op
+
+
+@pytest.fixture
+def caller_code(caller_gas, caller_type):
+    """
+    Code to call the callee contract:
+        PUSH1 0x00
+        DUP1 * 4
+        PUSH2 0xAAAA
+        PUSH2 call_gas
+        CALL/CALLCODE
+        PUSH1 0x00
+        SSTORE
+    """
+    return Op.SSTORE(0, caller_type(caller_gas, 0xAAAA, 0, 0, 0, 0, 0))
+
+
+@pytest.fixture
+def callee_code():
+    """
+    Code called by the caller contract:
+        PUSH1 0x00 * 4
+        PUSH1 0x00
+        PUSH1 0x00
+        PUSH2 0xAACC
+        PUSH2 0x1a90
+        CALLCODE
+    """
+    return Op.CALLCODE(0x1A90, 0xAACC, 1, 0, 0, 0, 0)
+
+
+@pytest.fixture
+def caller_tx() -> Transaction:
+    """
+    Transaction that calls the callee contract.
+    """
+    return Transaction(
+        chain_id=0x01,
+        nonce=0,
+        to=to_address(0xAAAB),
+        value=1,
+        gas_limit=500000,
+        gas_price=7,
+    )
+
+
+@pytest.fixture
+def pre(callee_code, caller_code) -> Dict[str, Account]:  # noqa: D103
+    return {
+        to_address(0xAAAA): Account(
+            balance=0x03,
+            code=callee_code,
+            nonce=1,
+        ),
+        to_address(0xAAAB): Account(
+            balance=0x03,
+            code=caller_code,
+            nonce=1,
+        ),
+        TestAddress: Account(
+            balance=0x0BA1A9CE,
+        ),
+    }
+
+
+@pytest.fixture
+def post(caller_gas) -> Dict[str, Account]:  # noqa: D103
+    return {
+        to_address(0xAAAB): Account(storage={0x00: (0x01 if caller_gas >= 0x3D50 else 0x00)}),
+    }
+
+
+@pytest.mark.parametrize(
+    "caller_gas",
+    [0x2D5F, 0x3D50],
```

Comment:
I'm just wondering why `0x2D5F` is insufficient and `0x3D50` is sufficient. Can this be made tighter? Or why is there such a jump between the two values. Increasing `0x2D5F` doesn't cause the test to fail; it would be nice if this was a hard limit. Thanks for explaining, perhaps I've missed something here!

Reply:
Glad you flagged this up - I assumed both CALL/CALLCODE would have the same insufficient gas values when refactoring. I checked in hive, and its not working properly. Apologies

From: https://github.com/ethereumjs/ethereumjs-monorepo/pull/3214/files#diff-ca2747b296607c2b77ee28d7067ae931408fd2e023d421b588bed952d9cb035fR729-R732
```
    ['f1', 36600, '0x'], // 36600 is CALL fee
    ['f2', 11600, '0x'], // 11600 is CALLCODE fee
    ['f1', 36600 + 7 * 3, '0x01'], // 36600 is CALL fee + 7 * 3 gas for 7 PUSH opcodes
    ['f2', 11600 + 7 * 3, '0x01'], // 11600 is CALLCODE fee + 7 * 3 gas for 7 PUSH opcodes
```
We should be using these values! I'll update this just now.

Reply:
Ah right, great find - yes especially the last two gas constants ensure that exactly enough gas is provided for CALL(CODE), such that that call frame does not run OOG